### ISSUE TITLE
case-blockscope? block scope for switch cases, something like that

### ISSUE THREAD

**Issue by joshxyzhimself** [eslint/eslint#14958]
**Please describe what the rule should do:**
Enforce curly-brace block-scopes in switch-case statements.

**What new ECMAScript feature does this rule relate to?**
No idea.

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)

**Provide 2-3 code examples that this rule will warn about:**

- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch#block-scope_variables_within_switch_statements

```js
const action = 'say_hello';
switch (action) {
  case 'say_hello':
    const message = 'hello';
    console.log(message);
    break;
  case 'say_hi':
    const message = 'hi';
    console.log(message);
    break;
  default:
    console.log('Empty action received.');
    break;
}

// output: Error: Identifier 'message' has already been declared
// eslint should warn or error here, recommend to use curly-braces for block-scopes
```

```js
const action = 'say_hello';
switch (action) {
  case 'say_hello': { // added curly-brace
    const message = 'hello';
    console.log(message);
    break;
  } // added curly-brace
  case 'say_hi': { // added curly-brace
    const message = 'hi';
    console.log(message);
    break;
  } // added curly-brace
  default: { // added curly-brace
    console.log('Empty action received.');
    break;
  } // added curly-brace
}

// output: "hello"
```

**Why should this rule be included in ESLint (instead of a plugin)?**
It's good practice like `default-case` and `no-fallthrough`, but that's for me. Do you guys think it should be?

**Are you willing to submit a pull request to implement this rule?**
if my iq allows it, why not :joy_cat:

**Comment by joshxyzhimself** [comment]
https://eslint.org/docs/rules/curly doesn't handle this, maybe add it there? wdyt?

**Comment by mdjermanovic** [comment]
Hi @joshxyzhimself, thanks for the issue!

> ```js
> const action = 'say_hello';
> switch (action) {
>   case 'say_hello':
>     const message = 'hello';
>     console.log(message);
>     break;
>   case 'say_hi':
>     const message = 'hi';
>     console.log(message);
>     break;
>   default:
>     console.log('Empty action received.');
>     break;
> }
> 
> // output: Error: Identifier 'message' has already been declared
> // eslint should warn or error here, recommend to use curly-braces for block-scopes
> ```

At this point, the code has a syntax error (`Identifier 'message' has already been declared` is a parsing error) so ESLint doesn't run rules and thus cannot recommend a fix.

However, assuming that the code isn't copy-pasted, and you start writing it as:

```js
const action = 'say_hello';
switch (action) {
  case 'say_hello':
    const message = 'hello';
    console.log(message);
    break;
}
```

the [no-case-declarations](https://eslint.org/docs/rules/no-case-declarations) rule will already report an error, as in [this online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLWNhc2UtZGVjbGFyYXRpb25zOiBlcnJvciAqL1xuXG5jb25zdCBhY3Rpb24gPSAnc2F5X2hlbGxvJztcbnN3aXRjaCAoYWN0aW9uKSB7XG4gIGNhc2UgJ3NheV9oZWxsbyc6XG4gICAgY29uc3QgbWVzc2FnZSA9ICdoZWxsbyc7XG4gICAgY29uc29sZS5sb2cobWVzc2FnZSk7XG4gICAgYnJlYWs7XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==). and the rule's documentation recommend the fix:

> To ensure that the lexical declaration only applies to the current case clause wrap your clauses in blocks

so I think this is already covered by the `no-case-declarations` rule, but we could consider adding a [suggestion](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions) if you think it would be useful.

**Comment by mdjermanovic** [comment]
No response so closing.
